common/src/models/error.rs

```diff
@@ -1,226 +1,171 @@
-use std::fmt::{Display, Formatter};
-
-use http::StatusCode;
+use http::{status::InvalidStatusCode, StatusCode};
 use serde::{Deserialize, Serialize};
 
-#[cfg(feature = "display")]
-use crossterm::style::Stylize;
-
 #[cfg(feature = "axum")]
 impl axum::response::IntoResponse for ApiError {
     fn into_response(self) -> axum::response::Response {
+        #[cfg(feature = "tracing-in-errors")]
         tracing::warn!("{}", self.message);
 
-        (self.status(), axum::Json(self)).into_response()
+        (
+            self.status().unwrap_or(StatusCode::INTERNAL_SERVER_ERROR),
+            axum::Json(self),
+        )
+            .into_response()
     }
 }
 
 #[derive(Serialize, Deserialize, Debug)]
 #[cfg_attr(feature = "utoipa", derive(utoipa::ToSchema))]
 #[typeshare::typeshare]
 pub struct ApiError {
-    pub message: String,
-    pub status_code: u16,
+    message: String,
+    status_code: u16,
 }
 
 impl ApiError {
-    pub fn internal(message: &str) -> Self {
+    #[inline(always)]
+    pub fn new(message: impl Into<String>, status_code: StatusCode) -> Self {
         Self {
-            message: message.to_string(),
-            status_code: StatusCode::INTERNAL_SERVER_ERROR.as_u16(),
+            message: message.into(),
+            status_code: status_code.as_u16(),
+        }
+    }
+    #[inline(always)]
+    pub fn status(&self) -> Result<StatusCode, InvalidStatusCode> {
+        StatusCode::from_u16(self.status_code)
+    }
+    #[inline(always)]
+    pub fn message(&self) -> &str {
+        self.message.as_str()
+    }
+
+    /// Create a one-off internal error with a string message exposed to the user.
+    #[inline(always)]
+    pub fn internal(message: impl AsRef<str>) -> Self {
+        #[cfg(feature = "tracing-in-errors")]
+        {
+            /// Dummy wrapper to allow logging a string `as &dyn std::error::Error`
+            #[derive(Debug)]
+            struct InternalError(String);
```

Comment:
@oddgrd I've assumed this helper is needed in order to make a string error correctly show up in DD through the `&dyn std::error::Error` path. Thoughts?

Reply:
Correct, we use a visitor in the tracing layer, and the type is how we know which field to record: https://docs.rs/tracing-core/latest/tracing_core/field/trait.Visit.html#method.record_error